lib/Differentiator/ActivityAnalyzer.cpp

```diff
@@ -0,0 +1,169 @@
+      for (const VarDecl* i : *m_BlockData[block.getBlockID()])
+        m_LoopMem.insert(i);
+    }
+
+    if (shouldPushSucc)
+      m_CFGQueue.insert(succ->getBlockID());
+
+    merge(succData.get(), m_BlockData[block.getBlockID()].get());
+  }
+  // FIXME: Information about the varied variables is stored in the last block,
+  // so we should be able to get it form there
+  for (const VarDecl* i : *m_BlockData[block.getBlockID()])
+    m_VariedDecls.insert(i);
+}
+
+bool VariedAnalyzer::isVaried(const VarDecl* VD) {
+  VarsData& curBranch = getCurBlockVarsData();
+  return curBranch.find(VD) != curBranch.end();
+}
+
+void VariedAnalyzer::merge(VarsData* targetData, VarsData* mergeData) {
+  for (const VarDecl* i : *mergeData)
+    targetData->insert(i);
+  for (const VarDecl* i : *targetData)
+    mergeData->insert(i);
+}
+
+void VariedAnalyzer::copyVarToCurBlock(const clang::VarDecl* VD) {
+  VarsData& curBranch = getCurBlockVarsData();
+  curBranch.insert(VD);
+}
+
+bool VariedAnalyzer::VisitBinaryOperator(BinaryOperator* BinOp) {
+  Expr* L = BinOp->getLHS();
+  Expr* R = BinOp->getRHS();
+
+  if (BinOp->isAssignmentOp()) {
+    m_Varied = false;
+    TraverseStmt(R);
+    m_Marking = m_Varied;
+    TraverseStmt(L);
+    m_Marking = false;
+  } else {
+    TraverseStmt(L);
+    TraverseStmt(R);
+  }
+  return true;
+}
+
+// add branching merging
+bool VariedAnalyzer::VisitConditionalOperator(ConditionalOperator* CO) {
+  TraverseStmt(CO->getCond());
+  TraverseStmt(CO->getTrueExpr());
+  TraverseStmt(CO->getFalseExpr());
+  return true;
+}
+
+bool VariedAnalyzer::VisitCallExpr(CallExpr* CE) {
+  FunctionDecl* FD = CE->getDirectCallee();
+  bool noHiddenParam = (CE->getNumArgs() == FD->getNumParams());
+  std::set<const clang::VarDecl*> variedParam;
+  if (noHiddenParam) {
+    MutableArrayRef<ParmVarDecl*> FDparam = FD->parameters();
+    for (std::size_t i = 0, e = CE->getNumArgs(); i != e; ++i) {
+      clang::Expr* par = CE->getArg(i);
+      TraverseStmt(par);
+      if (m_Varied) {
+        m_VariedDecls.insert(FDparam[i]);
+        m_Varied = false;
+      }
+    }
+  }
+  return true;
+}
+
+bool VariedAnalyzer::VisitDeclStmt(DeclStmt* DS) {
+  for (Decl* D : DS->decls()) {
+    if (auto* VD = dyn_cast<VarDecl>(D)) {
+      if (Expr* init = VD->getInit()) {
+        m_Varied = false;
+        TraverseStmt(init);
+        m_Marking = true;
+        VarsData& curBranch = getCurBlockVarsData();
+        if (m_Varied && curBranch.find(VD) == curBranch.end())
+          copyVarToCurBlock(VD);
+        m_Marking = false;
+      }
+    }
+  }
+  return true;
+}
+
+bool VariedAnalyzer::VisitUnaryOperator(UnaryOperator* UnOp) {
+  Expr* E = UnOp->getSubExpr();
+  TraverseStmt(E);
+  return true;
+}
+
+bool VariedAnalyzer::VisitDeclRefExpr(DeclRefExpr* DRE) {
+  if (isVaried(dyn_cast<VarDecl>(DRE->getDecl())))
+    m_Varied = true;
+
+  if (const auto* VD = dyn_cast<VarDecl>(DRE->getDecl())) {
+    VarsData& curBranch = getCurBlockVarsData();
```

Comment:
```suggestion
   const VarsData& curBranch = getCurBlockVarsData();
```

We should extract this into a utility function which does `getCurBlockVarsData` and `find`.